pkg/storage/tsdb/expanded_postings_cache.go

```diff
@@ -0,0 +1,422 @@
+	return c.fetchPostings(blockID, ix, ms...)(ctx)
+}
+
+func (c *BlocksPostingsForMatchersCache) fetchPostings(blockID ulid.ULID, ix tsdb.IndexReader, ms ...*labels.Matcher) func(context.Context) (index.Postings, error) {
+	var seed string
+	cache := c.blocksCache
+
+	// If is a head block, lets add the seed on the cache key so we can
+	// invalidate the cache when new series are created for this metric name
+	if isHeadBlock(blockID) {
+		cache = c.headCache
+		metricName, ok := metricNameFromMatcher(ms)
+		// Lets not cache head if we don;t find an equal matcher for the label __name__
+		if !ok {
+			return func(ctx context.Context) (index.Postings, error) {
+				return tsdb.PostingsForMatchers(ctx, ix, ms...)
+			}
+		}
+
+		seed = c.getSeedForMetricName(metricName)
+	}
+
+	// Let's bypass cache if not enabled
+	if !cache.cfg.Enabled {
+		return func(ctx context.Context) (index.Postings, error) {
+			return tsdb.PostingsForMatchers(ctx, ix, ms...)
+		}
+	}
+
+	c.metrics.CacheRequests.WithLabelValues(cache.name).Inc()
+
+	fetch := func() ([]storage.SeriesRef, int64, error) {
+		// Use context.Background() as this promise is maybe shared across calls
+		postings, err := c.postingsForMatchersFunc(context.Background(), ix, ms...)
+
+		if err == nil {
+			ids, err := index.ExpandPostings(postings)
+			return ids, int64(len(ids) * 8), err
+		}
+
+		return nil, 0, err
+	}
+
+	key := c.cacheKey(seed, blockID, ms...)
+	promise, loaded := cache.getPromiseForKey(key, fetch)
+	if loaded {
+		c.metrics.CacheHits.WithLabelValues(cache.name).Inc()
+	}
+
+	return c.result(promise)
+}
+
+func (c *BlocksPostingsForMatchersCache) result(promise *cacheEntryPromise[[]storage.SeriesRef]) func(ctx context.Context) (index.Postings, error) {
+	return func(ctx context.Context) (index.Postings, error) {
+		ids, err := promise.result(ctx)
+		return index.NewListPostings(ids), err
+	}
+}
+
+func (c *BlocksPostingsForMatchersCache) getSeedForMetricName(metricName string) string {
+	h := MemHashString(metricName)
+	i := h % uint64(len(c.headSeedByMetricName))
+	l := h % uint64(len(c.strippedLock))
+	c.strippedLock[l].RLock()
+	defer c.strippedLock[l].RUnlock()
+	return strconv.Itoa(c.headSeedByMetricName[i])
+}
+
+func (c *BlocksPostingsForMatchersCache) cacheKey(seed string, blockID ulid.ULID, ms ...*labels.Matcher) string {
+	slices.SortFunc(ms, func(i, j *labels.Matcher) int {
+		if i.Type != j.Type {
+			return int(i.Type - j.Type)
+		}
+		if i.Name != j.Name {
+			return strings.Compare(i.Name, j.Name)
+		}
+		if i.Value != j.Value {
+			return strings.Compare(i.Value, j.Value)
+		}
+		return 0
+	})
+
+	const (
+		typeLen = 2
+		sepLen  = 1
+	)
+
+	var size int
+	for _, m := range ms {
+		size += len(seed) + len(blockID.String()) + len(m.Name) + len(m.Value) + typeLen + 2*sepLen
+	}
+	sb := strings.Builder{}
+	sb.Grow(size)
+	sb.WriteString(seed)
+	sb.WriteByte('|')
```

Comment:
nit: should we skip `|` if the seed is an empty string?

Reply:
Maybe.. I think i would avoid a condition here just for readability tbh.